src/dune_engine/scheduler.mli

```diff
@@ -57,17 +57,27 @@ module Run : sig
       cases *)
   exception Shutdown_requested
 
-  (** Runs [once] in a loop, executing [finally] after every iteration, even if
-      Fiber.Never was encountered.
+  (** [poll once] runs [once] in a loop.
 
       If any source files change in the middle of iteration, it gets canceled.
 
       If [shutdown] is called, the current build will be canceled and new builds
       will not start. *)
   val poll :
-       (   report_error:(Exn_with_backtrace.t -> unit)
-        -> unit
-        -> [ `Continue | `Stop ] Fiber.t)
+       (report_error:(Exn_with_backtrace.t -> unit) -> [ `Continue ] Fiber.t)
```

Comment:
Could we make `poll` and `poll_passive` have a more similar API? I find it odd that one takes a "one step" function, while the other takes a `get_build_request` that returns a "one step" function. The fact that `get_build_request` is blocking on the RPC is an implementation detail that is not very interesting from the point of view the scheduler.

Reply:
I'm thinking of the RPC as not being the part of the build because including RPCs into build seems weird (breaks some intuitions about the build being deterministic).
But there are also practical reasons related to scheduler:
- The build can be cancelled by a filesystem event, while the wait for the RPC can not.
- The inotify sync needs to happen after the rpc, but before the build.

I'm not sure how to abstract that away.
Overall, I feel like polling loop is inherently going to interact with these concerns, but the scheduler does not have to. Having polling loop integrated with the scheduler is what seems weird to me.

I took a bite at one step in what seems the right direction to me in:
https://github.com/aalekseyev/dune/commit/dbf0c251fe06adface106927cd996a215e3dc8f7

This moves the filesystem event notifications into a separate variable (Invalidation_accumulator_var.t), but it's not really achieving enough to claim any separation.

Reply:
Yeah, I'm not arguing for having a single function, but at least similar APIs to learn to work with. The two `run_...` functions in `build_cmd.ml` end up having pretty different structure, which makes the code harder to review. If the two `poll` function had a similar API, then the two functions in `build_cmd.ml` should be more similar as well.

Reply:
Oh, got it. I'd argue `build_cmd.ml` is messy for a different reason: because I wrote it badly.
I just pushed a commit that simplifies code in `build_cmd.ml`.

As for the `scheduler.mli`, my initial desire was to move the entire implementation of `poll_passive` into `build_cmd.ml`, but I couldn't do it without exposing details of the state machine.  Any ideas for any specific improvements are very welcome!